gpflow/base.py

```diff
@@ -10,35 +10,74 @@
 
 from .config import default_float
 
+if TYPE_CHECKING:
+    from IPython.lib import pretty
+
 DType = Union[np.dtype, tf.DType]
-VariableData = Union[List, Tuple, np.ndarray, int, float]
+VariableData = Union[List, Tuple, np.ndarray, int, float]  # deprecated
 Transform = Union[tfp.bijectors.Bijector]
 Prior = Union[tfp.distributions.Distribution]
+UNKNOWN = Any
+
+S = TypeVar("S")
+_NestedSeq = Union[
+    Sequence[S],
+    Sequence[Sequence[S]],
+    Sequence[Sequence[Sequence[S]]],
+    Sequence[Sequence[Sequence[Sequence[Any]]]],  # catch all for rank > 3
+]
+_NativeScalar = Union[int, float]
+
+
+_Array = _NestedSeq[_NativeScalar]
+# represents a nested `Sequence` of `int` or `float`, such as
+#
+# 1, 1.1, [1.1], or [[1.1], [1, 1.1]]
+#
+# The ranks must be consistent, but the lengths are unconstrained
```

Comment:
I find this to be very hard-to-comprehend, I think it's reducing the readability of the source file significantly. I also think the type is wrong (but it's so hard to comprehend it's hard to be sure). What is the value we get out of it in return? I'd say use `Sequence[Union[int, float]]` and be done with it...

Reply:
> I find this to be very hard-to-comprehend, I think it's reducing the readability of the source file significantly.

ok. that's a concern. I personally find it makes it easier to read cos I then know what's meant to do what. I suppose my thought would be that it's easier to debug a mypy error than a runtime error. We're trading one for the probability of the other

> I also think the type is wrong (but it's so hard to comprehend it's hard to be sure).

I'd be interested to know in what way you think it's wrong, even just from an academic point of view

> What is the value we get out of it in return?

It's a bit hard to answer this in general. Most of the advantage is at call site: when someone uses the relevant methods with something that's not array-shaped, mypy would say

> I'd say use Sequence[Union[int, float]] and be done with it...

It would have to be `Sequence[Any]`. I could do that